### PR TITLE
Gradient centralization across all 4 CFD benchmarks

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Gradient Centralization (GC, Yong et al. 2020) subtracts the mean of each parameter's gradient tensor along all axes except the output-neuron axis before the optimizer step. This projects gradients onto a centered hyperplane, implicitly constraining weight space, reducing internal covariate shift, and acting as a form of weight regularization. GC has shown consistent improvements on image classification and NLP tasks with negligible compute overhead. In physics surrogate models with large, irregular meshes and high-variance pressure targets, noisy gradients may be a bottleneck — GC could reduce this without requiring any architectural changes.

**Cross-dataset scope:** TandemFoil · TandemFoil Paper · AirfRANS · DrivAerML

**Reference:** Yong et al., "Gradient Centralization: A New Optimization Technique for Deep Neural Networks", ECCV 2020.

---

## Instructions to the Student

Add a `--grad-centralization` boolean flag (default `False`) to `train.py`. When active, apply gradient centralization to all parameter gradients before the optimizer step. Insert a gradient hook or a post-backward pass loop:

```python
def centralize_gradient(grad, use_gc=True):
    """Subtract mean across all dims except the first (output neuron dim)."""
    if use_gc and grad.dim() > 1:
        # mean over all axes except axis 0
        grad = grad - grad.mean(dim=tuple(range(1, grad.dim())), keepdim=True)
    return grad

# Register on all parameters (excluding 1D, e.g. biases and norms)
for p in model.parameters():
    if p.dim() > 1:
        p.register_hook(lambda g: centralize_gradient(g))
```

This should be applied **after** loss.backward() and **before** optimizer.step(). Do not apply GC to 1D parameters (biases, LayerNorm/RMSNorm scale parameters).

Use `--wandb_group gradient-centralization` on all runs.

---

### TandemFoil

```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil \
  --optimizer lion --lr 1.25e-4 \
  --cosine-t-max 10 --grad-clip 0.5 --weight-decay 1e-2 \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame \
  --enable-cp-panel --enable-cp-panel-tandem-only \
  --asinh-pressure --residual-prediction --enable-pressure-prior-addition \
  --epochs 999 --ema-decay 0.999 \
  --grad-centralization \
  --wandb_group gradient-centralization
```

### TandemFoil Paper

```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil_paper \
  --optimizer adamw --lr 5e-4 \
  --cosine-t-max 150 --no-use-ema \
  --enable-fourier \
  --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --epochs 999 \
  --grad-centralization \
  --wandb_group gradient-centralization
```

### AirfRANS

```bash
cd target/icml2026 && python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 \
  --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --grad-centralization \
  --wandb_group gradient-centralization
```

### DrivAerML

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 --no-use-ema \
  --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --grad-centralization \
  --wandb_group gradient-centralization
```

---

## Current Baselines (lower is better)

| Dataset | Metric | Current Best | PR |
|---|---|---|---|
| TandemFoil | `val_primary/surface_pressure_mae` | 22.537 | #2924 |
| TandemFoil Paper | `val_primary/field_mse` | 0.00434 | #2979 |
| AirfRANS | `val_primary/surface_mse` | 0.000482 | #2951 |
| DrivAerML | `val_primary/surface_rel_l2_pct` | 3.997% | #2898 |

---

## Reporting

When done, post results as a PR comment with:
- W&B run IDs for all runs
- Best val metric per dataset (GC vs no-GC baseline)
- Whether GC beat baseline on each dataset
- Any notes on gradient norm behavior with vs without GC

### CODE REVIEW

**Comment by morganmcg1** [comment]
Closing to narrow the queue for the final ICML benchmark sprint. Current priority is benchmark-specific optimization and benchmark-faithful reporting for DrivAerML, TandemFoil paper benchmark, AirfRANS full Surf/Vol, and the Tandem parity anchor. If this idea still matters later, we can reopen it in a benchmark-specific form.
